pippy/PipelineStage.py

```diff
@@ -698,27 +868,73 @@ def forward(self, *args, **kwargs):
                 self.kwargs_chunk_spec,
             )
 
-        if self.rank == self.nstages - 1:
-            # where is my Y?
-            print(
-                f"[Rank{self.rank}] ArgsSplit: {args_split}, KwargsSplit: {kwargs_split}"
-            )
-
         # Activation send requests of all chunk
         all_send_reqs: List[dist.Work] = []
 
         output_chunks = [None] * self.chunks
 
-        # Forward pass of all chunks
-        for chunk in range(self.chunks):
-            s = self.streams[chunk % self.nstreams]
-            with torch.cuda.stream(s):
-                output, send_reqs = self.forward_one_chunk(
-                    chunk, args_split, kwargs_split, fwd_cache
-                )
-                all_send_reqs += send_reqs
-                # Prepare for final output merge or reduction
-                output_chunks[chunk] = output
+        if self.inner_depth > 1:
+            if self.enable_efficient_inner_pipe:
+                # New schedule:
+                # 1A, 2A, 1M, 2M
+                for i in range(self.inner_depth):
+                    for chunk in range(self.chunks):
+                        # each stream maintains dependency so it is shared by same chunk computation ops
+                        s = self.streams[chunk % self.nstreams]
+                        with torch.cuda.stream(s):
+                            if chunk == 0:
+                                (
+                                    composite_args,
+                                    composite_kwargs,
+                                ) = self._recv_and_fill_inputs(
+                                    chunk,
+                                    args_split,
+                                    kwargs_split,
+                                )
+                            (
+                                output,
+                                send_reqs,
+                            ) = self.forward_one_chunk_one_ipipe(
+                                chunk,
+                                i,
+                                args_split,
+                                kwargs_split,
+                                composite_args,
+                                composite_kwargs,
+                                fwd_cache,
+                            )
+                            if i == self.inner_depth - 1:
+                                assert (
+                                    output is not None and send_reqs is not None
+                                )
+                                all_send_reqs += send_reqs
+                                # Prepare for final output merge or reduction
+                                output_chunks[chunk] = output
+
+            else:
+                # Forward pass of all chunks
```

Comment:
The comment here is same as L928. Can we make this more clear and easy to understand that each condition is doing?